modulemd/modulemd-module-stream-v2.c

```diff
@@ -734,6 +734,47 @@ modulemd_module_stream_v2_get_profile (ModulemdModuleStreamV2 *self,
 }
 
 
+struct profile_match_ctx
+{
+  GHashTable *profiles;
+  GPtrArray *found;
+  const gchar *pattern;
+};
+
+static void
+profile_match (gpointer data, gpointer user_data)
+{
+  struct profile_match_ctx *match_ctx = (struct profile_match_ctx *)user_data;
+
+  /* Add it to the found list if it matches the pattern */
+  if (modulemd_fnmatch (match_ctx->pattern, (const gchar *)data))
+    {
+      g_ptr_array_add (match_ctx->found,
+                       g_object_ref (g_hash_table_lookup (
+                         match_ctx->profiles, (const gchar *)data)));
+    }
+}
+
+GPtrArray *
+modulemd_module_stream_v2_search_profiles (ModulemdModuleStreamV2 *self,
+                                           const gchar *profile_pattern)
+{
+  /* We will commonly only match a single profile, so allocate for that */
+  GPtrArray *found = g_ptr_array_new_full (1, g_object_unref);
+
```

Comment:
> I[t] could be argued if the common use case will match only a single profile, or all profiles.

Hmm, given that it's unlikely for the number of profiles to grow very large, you're probably right that it's better to assume it matches all of them.

> Also, might it be a good idea to add `g_return_val_if_fail (MODULEMD_IS_MODULE_STREAM_V2 (self), found);` here? Or is it better to let it (hopefully) blow up with an exception?

No, you're right. I forgot to add it. Since the function description says "this function cannot fail", we should return the empty list on a programming error. (It will assert in debug builds.)